mjpc/planners/sample_gradient/planner.cc

```diff
@@ -0,0 +1,652 @@
+}
+
+// rollout candidate policies
+void SampleGradientPlanner::Rollouts(int num_trajectory, int num_gradient,
+                                     int horizon, ThreadPool& pool) {
+  // reset perturbation compute time
+  noise_compute_time = 0.0;
+
+  // search
+  int count_before = pool.GetCount();
+  for (int i = 0; i < num_trajectory; i++) {
+    pool.Schedule([&s = *this, &model = this->model, &task = this->task,
+                   &state = this->state, &time = this->time,
+                   &mocap = this->mocap, &userdata = this->userdata, horizon,
+                   idx_nominal = this->idx_nominal, num_trajectory,
+                   num_gradient, i]() {
+      // nominal and noisy policies
+      if (i < num_trajectory - num_gradient) {
+        // copy nominal policy
+        s.candidate_policy[i].CopyFrom(s.resampled_policy,
+                                       s.resampled_policy.num_spline_points);
+        s.candidate_policy[i].representation =
+            s.resampled_policy.representation;
+
+        // noisy nominal policy
+        if (i > idx_nominal) s.AddNoiseToPolicy(i);
+      }
+
+      // ----- rollout sample policy ----- //
+
+      // policy
+      auto sample_policy_i = [&candidate_policy = s.candidate_policy, &i](
+                                 double* action, const double* state,
+                                 double time) {
+        candidate_policy[i].Action(action, state, time);
+      };
+
+      // policy rollout
+      s.trajectory[i].Rollout(
+          sample_policy_i, task, model, s.data_[ThreadPool::WorkerId()].get(),
+          state.data(), time, mocap.data(), userdata.data(), horizon);
+    });
+  }
+  pool.WaitCount(count_before + num_trajectory);
+  pool.ResetCount();
+}
+
+// compute candidate trajectories
+void SampleGradientPlanner::GradientCandidates(int num_trajectory,
+                                               int num_gradient, int horizon,
+                                               ThreadPool& pool) {
+  if (num_gradient < 1) return;
+
+  // number of parameters
+  int num_parameters = resampled_policy.num_parameters;
+  int num_spline_points = resampled_policy.num_spline_points;
+
+  // cache old gradient
+  mju_copy(gradient_previous.data(), gradient.data(), num_parameters);
+
+  // -- compute approximate gradient -- //
+  // average return
+  int num_noisy = num_trajectory - num_gradient;
+  double avg_return = 0.0;
+  for (int i = 0; i < num_noisy; i++) {
+    avg_return += trajectory[i].total_return;
+  }
+  avg_return /= num_noisy;
+
+  // gradient
+  std::fill(gradient.begin(), gradient.end(), 0.0);
+  for (int i = 1; i < num_noisy; i++) {
+    double* noisei = noise.data() + i * (model->nu * kMaxTrajectoryHorizon);
+    mju_addToScl(gradient.data(), noisei,
+                 (trajectory[i].total_return - avg_return) / (num_noisy - 1),
+                 num_parameters);
+  }
+
+  // normalize gradient
+  // TODO(taylor): should we normalize?
```

Comment:
yes however...

Reply:
if we ensure the returns are scaled correctly, we shouldn't need to normalize the gradient

Reply:
Removed normalization.